google/gax/__init__.py

```diff
@@ -482,3 +496,180 @@ def __next__(self):
+        self._get_operation()
+        return (self._operation.HasField('error') and
+                self._operation.error.code == code_pb2.CANCELLED)
+
+    def done(self):
+        """Issues OperationsClient.get_operation and returns value of
+           Operation.done.
+        """
+        return self._get_operation().done
+
+    def add_done_callback(self, fn):  # pylint: disable=invalid-name
+        """Enters a polling loop on OperationsClient.get_operation, and once the
+           operation is done or cancelled, calls the function with this
+           _OperationFuture. Added callables are called in the order that they
+           were added.
+        """
+        self._done_clbks.append(fn)
+        if self._process is None:
+            self._process = mp.Process(target=self._execute_clbks)
+            self._process.start()
+        elif not self._process.is_alive() and self._operation.done:
+            _try_callback(self, fn)
+
+    def operation_name(self):
+        """Returns the value of Operation.name from the last call to
+           OperationsClient.get_operation (or if only the initial API call has
+           been made, the name from that first call).
+        """
+        return self._operation.name
+
+    def metadata(self):
+        """Returns the value of Operation.metadata from the last call to
+           OperationsClient.get_operation (or if only the initial API call has
+           been made, the metadata from that first call).
+        """
+        if self._operation.metadata is None:
+            return None
+
+        return _from_any(self._metadata_type, self._operation.metadata)
+
+    def last_operation_data(self):
+        """Returns the data from the last call to OperationsClient.get_operation
+           (or if only the initial API call has been made, the data from that
+           first call).
+        """
+        return self._operation
+
+    def _get_operation(self):
+        if not self._operation.done:
+            self._operation = self._client.get_operation(
+                self._operation.name, self._call_options)
+
+        return self._operation
+
+    def _done_check(self, _timeout):
+        if self.done():
+            return self._operation
+
+        raise _DeadlineExceededError()
+
+    def _poll(self, timeout=None):
+        if timeout is None:
+            backoff_settings = BackoffSettings(
+                1000, 2, 30000, None, None, None, None)
+        else:
+            backoff_settings = BackoffSettings(
+                1000, 2, 30000, 0, 0, 0, timeout * _MILLIS_PER_SEC)
+
+        retry_options = RetryOptions(
+            [StatusCode.DEADLINE_EXCEEDED], backoff_settings)
+        retryable_done_check = retryable(self._done_check, retry_options)
+
+        return retryable_done_check()
+
+    def _execute_clbks(self):
+        self._poll()
+
+        for done_clbk in self._done_clbks:
+            _try_callback(self, done_clbk)
```

Comment:
Is there a reason why the callbacks are still being stored after the they are run?